flink-table-store-core/src/main/java/org/apache/flink/table/store/table/system/AuditLogTable.java

```diff
@@ -0,0 +1,282 @@
+            dataScan.withLevel(level);
+            return this;
+        }
+
+        @Override
+        public DataTableScan.DataFilePlan plan() {
+            return dataScan.plan();
+        }
+    }
+
+    private class AuditLogRead implements TableRead {
+
+        private final TableRead dataRead;
+
+        private int[] readProjection;
+
+        private AuditLogRead(TableRead dataRead) {
+            this.dataRead = dataRead;
+            this.readProjection = defaultProjection();
+        }
+
+        /** Default projection, just add row kind to the first. */
+        private int[] defaultProjection() {
+            int dataFieldCount = dataTable.rowType().getFieldCount();
+            int[] projection = new int[dataFieldCount + 1];
+            projection[0] = -1;
+            for (int i = 0; i < dataFieldCount; i++) {
+                projection[i + 1] = i;
+            }
+            return projection;
+        }
+
+        @Override
+        public TableRead withFilter(Predicate predicate) {
+            convert(predicate).ifPresent(dataRead::withFilter);
+            return this;
+        }
+
+        @Override
+        public TableRead withProjection(int[][] projection) {
+            // data projection to push down to dataRead
+            List<int[]> dataProjection = new ArrayList<>();
+            // read projection to handle record returned by dataRead
+            List<Integer> readProjection = new ArrayList<>();
+            boolean rowKindAppeared = false;
+            for (int i = 0; i < projection.length; i++) {
+                int[] field = projection[i];
+                int topField = field[0];
+                if (topField == 0) {
+                    rowKindAppeared = true;
+                    readProjection.add(-1);
+                } else {
+                    int[] newField = Arrays.copyOf(field, field.length);
+                    newField[0] = newField[0] - 1;
+                    dataProjection.add(newField);
+
+                    // There is no row kind field. Keep it as it is
+                    // Row kind field has occurred, and the following fields are offset by 1
+                    // position
+                    readProjection.add(rowKindAppeared ? i - 1 : i);
+                }
+            }
+            this.readProjection = Ints.toArray(readProjection);
+            dataRead.withProjection(dataProjection.toArray(new int[0][]));
+            return this;
+        }
+
+        @Override
+        public RecordReader<RowData> createReader(Split split) throws IOException {
+            return transform(dataRead.createReader(split), this::convertRow);
+        }
+
+        private RowData convertRow(RowData data) {
+            return new AuditLogRowData(readProjection, data);
+        }
+    }
+
+    /** A {@link MappingRowData} which returns row kind when mapping index is negative. */
+    private static class AuditLogRowData extends MappingRowData {
```

Comment:
No need to specially introduce a `MappingRowData`. Just keep the old `ProjectedRowData` and let this class extend `ProjectedRowData`.